test/unit/ADT/ClonePointer.cpp

```diff
@@ -0,0 +1,89 @@
+
+#include "caffeine/ADT/ClonePointer.h"
+#include <gtest/gtest.h>
+
+#include <vector>
+
+using namespace caffeine;
+
+class Basic {
+public:
+  int a = 0;
+  Basic(int b) : a{b} {};
+  virtual std::unique_ptr<Basic> clone() {
+    return std::make_unique<Basic>(a);
+  }
+
+  virtual int get_i() {
+    return a;
+  }
+
+  virtual ~Basic() = default;
+};
+
+TEST(clone_ptr, can_construct) {
+  clone_ptr<Basic> basic = make_clone<Basic>(5);
+  ASSERT_EQ(basic->a, 5);
+}
+
+TEST(clone_ptr, copy_does_copy) {
+  clone_ptr<Basic> basic = make_clone<Basic>(5);
+  ASSERT_EQ(basic->a, 5);
+  auto basic_2 = basic;
+  ASSERT_EQ(basic_2->a, 5);
+  ASSERT_EQ(basic->a, 5);
+  basic->a = 6;
+  ASSERT_EQ(basic_2->a, 5);
+  ASSERT_EQ(basic->a, 6);
+}
+
+class Basic5 : public Basic {
+public:
+  Basic5() : Basic(0){};
+  int get_i() override {
+    return 5;
+  }
+
+  std::unique_ptr<Basic> clone() override {
+    return std::make_unique<Basic5>();
```

Comment:
Shouldn't this be actually making a copy of the current class?
```suggestion
    return std::make_unique<Basic5>(*this);
```